src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,416 @@
+-- of adding a larger-than-page value to a run.
+--
+-- Note: the string of bytes is not padded to the target disk-page size.
+data RawPages = RawPages {
+      -- | The number of pages that the output bytestring spans
+      rawPageCount :: !Int
+    , rawPageBytes :: !BS.ByteString
+    }
+
+--
+-- Keys
+--
+
+-- | A string of raw bytes representing a key.
+type RawKey = S.SerialisedKey
+
+-- | Size of key in number of bytes.
+sizeofKey :: RawKey -> Int
+sizeofKey (S.SerialisedKey ba) = sizeofByteArray ba
+
+-- | Size of key in number of bytes.
+sizeofKey16 :: RawKey -> Word16
+sizeofKey16 = fromIntegral . sizeofKey
+
+-- | Size of key in number of bytes.
+sizeofKey64 :: RawKey -> Word64
+sizeofKey64 = fromIntegral . sizeofKey
+
+--
+-- Values
+--
+
+-- | A string of raw bytes representing a value.
+--
+-- A consequence of this simple representation is that for values spanning
+-- multiple pages, those pages must be read into contiguous, page-sized buffers.
+-- This might require dynamically growing the pool of buffers.
+data RawValue =
+    UnsafeRawValue !Int !Int ByteArray
+  | EmptyRawValue
+
+{-# COMPLETE EmptyRawValue, RawValue #-}
+pattern RawValue :: Int -> Int -> ByteArray -> RawValue
+pattern RawValue start end ba <- UnsafeRawValue start end ba
+  where RawValue start end ba =
+          assert (0 <= start && start <= end && end <= sizeofByteArray ba) $
+          UnsafeRawValue start end ba
+
+-- | Size of value in number of bytes.
+sizeofValue :: RawValue -> Int
+sizeofValue EmptyRawValue = 0
+sizeofValue (RawValue start end ba) =
+  max 0 $ min (end - start) (sizeofByteArray ba)
+
+-- | Size of value in number of bytes.
+sizeofValue16 :: RawValue -> Word16
+sizeofValue16 = fromIntegral . sizeofValue
+
+-- | Size of value in number of bytes.
+sizeofValue32 :: RawValue -> Word32
+sizeofValue32 = fromIntegral . sizeofValue
+
+-- | Size of value in number of bytes.
+sizeofValue64 :: RawValue -> Word64
+sizeofValue64 = fromIntegral . sizeofValue
+
+emptyValue :: RawValue
+emptyValue = EmptyRawValue
+
+--
+-- Operation
+--
+
+data Operation =
+    Insert RawValue
+  | Mupsert RawValue
+  | Delete
+
+toValue :: Operation -> RawValue
+toValue (Insert v)  = v
+toValue (Mupsert v) = v
+toValue Delete      = emptyValue
+
+opToCrumb :: Operation -> Crumb
+opToCrumb Insert{}  = 0
+opToCrumb Mupsert{} = 1
+opToCrumb Delete{}  = 2
+
+--
+-- Blob reference
+--
+
+-- | TODO: replace this placeholder type by the actual blob reference type.
+data BlobRef = BlobRef {
+    blobRefOffset :: !Word64
+  , blobRefSize   :: !Word32
+  }
+
+{-------------------------------------------------------------------------------
+  Page serialisation
+-------------------------------------------------------------------------------}
+
+serialisePage :: PageIntermediate -> RawPages
+serialisePage PageIntermediate{..} =
+      RawPages (fromIntegral $ pageSizeNumBytes `div` 4096)
+    . BS.toStrict
+    . BB.toLazyByteString $
```

Comment:
This implementation is temporary: it's used to show functional equivalence to the the `FormatPage` prototype.

Optimisation: going through a `Builder` and `LazyByteString` to end up with a `StrictByteString` is not ideal, because it means we are going to copy all the bytes from the `LazyByteString` to the `StrictByteString`. We can optimise this because we know exactly how many bytes of data we have accumulated in `PageIntermediate`, and we can allocate/re-use buffers accordingly. Note that if we want to put the raw page bytes into reusable buffers, then we might have to grow the pool of buffers dynamically. Furthermore, if these buffers are not contiguous, then `serialisePage` has to be aware of that too

Reply:
Option: use a single buffer and run the builder for multiple buffer-sized chunks